domain/agentbinary/service/controllerstore.go

```diff
@@ -0,0 +1,205 @@
+	// The following errors can be returned:
+	// - [agentbinaryerrors.ObjectNotFound] when no object exists that matches this path.
+	GetObjectUUID(ctx context.Context, path string) (objectstore.UUID, error)
+
+	// RegisterAgentBinary registers a new agent binary's metadata to the database.
+	// [agentbinaryerrors.AlreadyExists] when the provided agent binary already
+	// exists.
+	// [agentbinaryerrors.ObjectNotFound] when no object exists that matches
+	// this agent binary.
+	// [coreerrors.NotSupported] if the architecture is not supported by the
+	// state layer.
+	RegisterAgentBinary(ctx context.Context, arg agentbinary.RegisterAgentBinaryArg) error
+
+	// GetAgentBinarySHA256 checks if an agent binary record for the given
+	// version exists and, if found, returns its associated SHA256
+	// value. Note that this only verifies existence of a recorded agent binary entry,
+	// the underlying object may exist in the store without this association, in which case
+	// the function will return false.
+	GetAgentBinarySHA256(ctx context.Context, ver coreagentbinary.Version, stream coreagentbinary.AgentStream) (bool, string, error)
+}
+
+type ControllerAgentBinaryStore struct {
+	state             ControllerStoreState
+	logger            logger.Logger
+	objectStoreGetter objectstore.NamespacedObjectStoreGetter
+}
+
+// NewControllerAgentBinaryStore returns a new instance of ControllerAgentBinaryStore.
+func NewControllerAgentBinaryStore(
+	state ControllerStoreState,
+	logger logger.Logger,
+	objectStoreGetter objectstore.NamespacedObjectStoreGetter,
+) *ControllerAgentBinaryStore {
+	return &ControllerAgentBinaryStore{
+		state:             state,
+		logger:            logger,
+		objectStoreGetter: objectStoreGetter,
+	}
+}
+
+// GetAgentBinary retrieves the agent binary corresponding to the given version
+// and stream from the controller's object store.
+//
+// The function first queries controller state to check whether a SHA256 record
+// exists for the requested version and stream. If no such record is found,
+// agentbinaryerrors.NotFound is returned.
+//
+// If a valid SHA256 is recorded, the corresponding binary blob is fetched from
+// the controller's object store and returned as an io.ReadCloser together with
+// its size and SHA256 string.
+//
+// The caller is responsible for closing the returned reader.
+func (s *ControllerAgentBinaryStore) GetAgentBinary(ctx context.Context, ver coreagentbinary.Version, stream coreagentbinary.AgentStream) (io.ReadCloser, int64, string, error) {
+	hasAgentBinary, sha256Str, err := s.state.GetAgentBinarySHA256(ctx, ver, stream)
+	if err != nil {
+		return nil, 0, "", err
+	}
+	if !hasAgentBinary {
+		return nil, 0, "", agentbinaryerrors.NotFound
+	}
+
+	store, err := s.objectStoreGetter.GetObjectStore(ctx)
+	if err != nil {
+		return nil, 0, "", err
+	}
+	reader, size, err := store.GetBySHA256(ctx, sha256Str)
+	if err != nil {
+		return nil, 0, "", err
+	}
```

Comment:
Happy for this to happen in a follow up PR:

We want to transalte the objectstoreerrors.ObjectNotFound error. The reason for this is we need to maintain a contract with the caller. We say that to implement this interface if something isn't found it should be the `NotFound` error of the domain.

That way we know to move on try the next thing.

```suggestion
    if errors.Is(err, objectstoreerrors.ObjectNotFound) {
        return nil, 0, "", errors.New("agent binary not found in controller store").Add(agentbinaryerrors.NotFound)
    } else if {
        return nil, 0, "", errors.Errorf("getting agent binary with sha %q from controller object store: %w", sha256Str, err)
    }
```

Reply:
hi tom, just curious on this. If there were other nonfound errors, cant we move on to try the next thing as well? i.e if there is an err getting from state due to a transient db connection issue, would it make sense to try and get from simple streams instead?

Reply:
you could definitely do that but you are operating without context. What happens if the next operation returns an error? Do you aggregate the errors together or do you return the first or second one. What happens if going further does something we don't want. How is the caller suppose to convey this information on wards.

What I am trying to get at is it is only when we understand the failure case should we continue on. This is why documenting contracts is important as they give us reasonable expectations and assertions that can be relied upon.

Reply:
I see, that makes sense, thanks for the clarification tom. Initially my idea was that in this scenario we could just return the second but thinking more about it, it's not a great idea to just drop the first err and proceed silently. Have refactored this as well.